File: Resource.f90
program ResourceEvolution
    implicit none

    ! 集合論：状態空間と戦略空間
    integer, parameter :: n_strategies = 2  ! 戦略数（Cooperative, Defective）
    real(8), dimension(n_strategies) :: strategy_distribution  ! 戦略分布 x
    real(8) :: resource  ! リソース量 r
    real(8) :: fitness  ! 平均適応度 \bar{f}
    real(8), dimension(n_strategies) :: strategy_costs = (/0.3, 0.1/)  ! 戦略のコスト（環論）
    real(8), parameter :: max_resource = 10.0  ! 最大リソース量
    real(8), parameter :: fitness_threshold = 1.0  ! 適応度閾値
    real(8), parameter :: resource_decay = 0.1  ! リソース消費率
    real(8), parameter :: dt = 0.01  ! 時間ステップ（確率論）
    integer, parameter :: max_steps = 1000  ! 最大ステップ数

    ! 作業変数
    integer :: i, step
    real(8) :: final_fitness, final_resource
    logical :: high_fitness_reached

    ! 初期状態（集合論：状態空間の初期化）
    call initialize_state(strategy_distribution, resource, fitness)

    ! シミュレーション（圏論：状態遷移）
    call simulate_evolution(strategy_distribution, resource, fitness, &
                            final_fitness, final_resource, high_fitness_reached)

    ! 結果出力（モーダル論理：\diamond Fitness(s^*, r')）
    print *, 'Final Fitness:', final_fitness
    print *, 'Final Resource:', final_resource
    if (high_fitness_reached) then
        print *, 'Evolved to high fitness'
    else
        print *, 'Did not evolve to high fitness'
    end if

contains

    ! 初期状態の設定
    subroutine initialize_state(dist, res, fit)
        real(8), dimension(:), intent(out) :: dist
        real(8), intent(out) :: res, fit
        real(8) :: sum_dist

        ! 初期戦略分布（例：均等）
        dist(1) = 0.5  ! Cooperative
        dist(2) = 0.5  ! Defective
        sum_dist = sum(dist)
        dist = dist / sum_dist  ! 正規化（群論：対称性）

        ! 初期リソース（豊富）
        res = max_resource

        ! 初期適応度
        fit = 0.0
    end subroutine initialize_state

    ! 適応度計算（環論：実数環での評価）
    function compute_fitness(strategy_idx, res) result(f)
        integer, intent(in) :: strategy_idx
        real(8), intent(in) :: res
        real(8) :: f
        ! f(\sigma, r) \propto r / (1 + c(\sigma)) （体論：一貫性）
        f = res / (1.0 + strategy_costs(strategy_idx))
    end function compute_fitness

    ! 平均適応度計算
    function compute_average_fitness(dist, res) result(avg_f)
        real(8), dimension(:), intent(in) :: dist
        real(8), intent(in) :: res
        real(8) :: avg_f
        integer :: i
        avg_f = 0.0
        do i = 1, n_strategies
            avg_f = avg_f + dist(i) * compute_fitness(i, res)
        end do
    end function compute_average_fitness

    ! レプリケーターダイナミクス（確率論：レプリケーターダイナミクス）
    subroutine update_strategy_distribution(dist, res, fit)
        real(8), dimension(:), intent(inout) :: dist
        real(8), intent(in) :: res
        real(8), intent(out) :: fit
        real(8), dimension(n_strategies) :: new_dist
        real(8) :: avg_f, sum_dist
        integer :: i

        avg_f = compute_average_fitness(dist, res)
        sum_dist = 0.0
        do i = 1, n_strategies
            ! \dot{x}_\sigma = x_\sigma (f(\sigma, r) - \bar{f})
            new_dist(i) = dist(i) * (compute_fitness(i, res) - avg_f) * dt + dist(i)
            if (new_dist(i) < 0.0) new_dist(i) = 0.0
            sum_dist = sum_dist + new_dist(i)
        end do

        ! 正規化（群論：戦略分布の対称性）
        if (sum_dist > 0.0) then
            new_dist = new_dist / sum_dist
        end if
        dist = new_dist
        fit = compute_average_fitness(dist, res)
    end subroutine update_strategy_distribution

    ! 進化シミュレーション（圏論：状態遷移プロセス）
    subroutine simulate_evolution(dist, res, fit, final_fit, final_res, reached)
        real(8), dimension(:), intent(inout) :: dist
        real(8), intent(inout) :: res
        real(8), intent(inout) :: fit
        real(8), intent(out) :: final_fit, final_res
        logical, intent(out) :: reached
        real(8) :: avg_cost
        integer :: step

        final_fit = fit
        final_res = res
        reached = .false.

        do step = 1, max_steps
            if (res <= 0.0) exit

            ! リソース消費（環論：リソースの減法）
            avg_cost = sum(dist * strategy_costs)
            res = res - resource_decay * avg_cost
            if (res < 0.0) res = 0.0

            ! 戦略分布更新（確率論）
            call update_strategy_distribution(dist, res, fit)

            ! 適応度チェック（数理論理：HighFitness）
            if (fit >= fitness_threshold) then
                reached = .true.
                exit
            end if
        end do

        final_fit = fit
        final_res = res
    end subroutine simulate_evolution

end program ResourceEvolution

File: Resource.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ResourceEvolutionModel
{
    // 状態（戦略分布とリソース状態）
    public class State
    {
        public string Id { get; }
        public double[] StrategyDistribution { get; set; } // 戦略分布 x
        public double Resource { get; set; } // リソース量
        public double Fitness { get; set; } // 平均適応度
        public State(string id, double[] strategyDistribution, double resource)
        {
            Id = id;
            StrategyDistribution = strategyDistribution;
            Resource = resource;
            Fitness = 0.0;
        }
    }

    // 戦略
    public class Strategy
    {
        public string Name { get; }
        public double Cost { get; } // 戦略のコスト
        public Strategy(string name, double cost)
        {
            Name = name;
            Cost = cost;
        }
    }

    // アクセシビリティ関係（圏論的射 + 確率）
    public class Transition
    {
        public State Source { get; }
        public State Target { get; }
        public Strategy Strategy { get; }
        public double Probability { get; } // 遷移確率
        public Transition(State source, State target, Strategy strategy, double probability)
        {
            Source = source;
            Target = target;
            Strategy = strategy;
            Probability = probability;
        }
    }

    // 進化ゲームモデル
    public class ResourceEvolutionModel
    {
        private List<State> States { get; } // 状態空間
        private List<Strategy> Strategies { get; } // 戦略空間
        private List<Transition> Transitions { get; } // アクセシビリティ関係
        private readonly Random random;
        private readonly double dt = 0.01; // 時間ステップ
        private readonly double resourceDecay = 0.1; // リソース消費率

        public ResourceEvolutionModel()
        {
            States = new List<State>();
            Strategies = new List<Strategy>();
            Transitions = new List<Transition>();
            random = new Random();
        }

        // 状態の追加
        public void AddState(string id, double[] strategyDistribution, double resource)
        {
            States.Add(new State(id, strategyDistribution, resource));
        }

        // 戦略の追加
        public void AddStrategy(string name, double cost)
        {
            Strategies.Add(new Strategy(name, cost));
        }

        // アクセシビリティ関係の追加
        public void AddTransition(string sourceId, string targetId, string strategyName, double probability)
        {
            var source = States.Find(s => s.Id == sourceId);
            var target = States.Find(s => s.Id == targetId);
            var strategy = Strategies.Find(s => s.Name == strategyName);
            if (source != null && target != null && strategy != null)
            {
                Transitions.Add(new Transition(source, target, strategy, probability));
            }
        }

        // 適応度計算
        private double ComputeFitness(Strategy strategy, State state)
        {
            // 適応度はリソースに比例、コストに反比例
            return state.Resource / (1.0 + strategy.Cost);
        }

        // レプリケーターダイナミクス
        private void UpdateStrategyDistribution(State state)
        {
            double[] x = state.StrategyDistribution;
            double[] fitness = new double[Strategies.Count];
            double avgFitness = 0.0;

            // 各戦略の適応度
            for (int i = 0; i < Strategies.Count; i++)
            {
                fitness[i] = ComputeFitness(Strategies[i], state);
                avgFitness += x[i] * fitness[i];
            }

            // レプリケーターダイナミクス
            double[] newX = new double[x.Length];
            double sum = 0.0;
            for (int i = 0; i < x.Length; i++)
            {
                newX[i] = x[i] * (fitness[i] - avgFitness) * dt + x[i];
                if (newX[i] < 0) newX[i] = 0;
                sum += newX[i];
            }

            // 正規化
            for (int i = 0; i < x.Length; i++)
            {
                newX[i] /= sum;
            }

            state.StrategyDistribution = newX;
            state.Fitness = avgFitness;
        }

        // モーダル論理：高適応度状態への到達可能性（S4）
        public bool CanReachHighFitness(State state, double fitnessThreshold, int maxSteps = 100)
        {
            var visited = new HashSet<State>();
            var queue = new Queue<State>();
            queue.Enqueue(state);
            visited.Add(state);

            if (state.Fitness >= fitnessThreshold)
                return true;

            while (queue.Count > 0 && maxSteps-- > 0)
            {
                var current = queue.Dequeue();
                var accessible = Transitions.Where(t => t.Source == current);
                foreach (var transition in accessible)
                {
                    var next = transition.Target;
                    if (next.Fitness >= fitnessThreshold)
                        return true;
                    if (!visited.Contains(next))
                    {
                        queue.Enqueue(next);
                        visited.Add(next);
                    }
                }
            }
            return false;
        }

        // 進化シミュレーション
        public (double, double) SimulateEvolution(State initial, int steps)
        {
            var current = initial;
            double finalFitness = current.Fitness;
            double finalResource = current.Resource;

            for (int t = 0; t < steps; t++)
            {
                var transitions = Transitions.Where(t => t.Source == current).ToList();
                if (!transitions.Any())
                    break;

                // リソース消費
                finalResource -= resourceDecay * transitions.Average(t => t.Strategy.Cost);
                if (finalResource < 0) finalResource = 0;
                current.Resource = finalResource;

                // 戦略分布更新
                UpdateStrategyDistribution(current);

                // 状態遷移
                double totalProb = transitions.Sum(t => t.Probability);
                double r = random.NextDouble() * totalProb;
                double sum = 0;
                foreach (var t in transitions)
                {
                    sum += t.Probability;
                    if (r <= sum)
                    {
                        current = t.Target;
                        break;
                    }
                }

                finalFitness = current.Fitness;
                if (finalFitness >= 1.0) // 高適応度閾値
                    break;
            }

            return (finalFitness, finalResource);
        }

        // モーダル論理評価：\diamond Fitness(s^*, r')
        public bool EvaluateProposition(State initialState, double fitnessThreshold = 1.0)
        {
            // モーダル論理：高適応度状態への到達可能性
            bool isHighFitnessPossible = CanReachHighFitness(initialState, fitnessThreshold);
            if (!isHighFitnessPossible)
                return false;

            // 確率論的評価
            var (finalFitness, finalResource) = SimulateEvolution(initialState, 1000);
            return finalFitness >= fitnessThreshold && finalResource >= 0;
        }

        // シミュレーション実行
        public void RunSimulation()
        {
            foreach (var state in States)
            {
                bool result = EvaluateProposition(state);
                Console.WriteLine($"State {state.Id} (Resource={state.Resource}): {(result ? "Evolves to high fitness" : "Does not evolve to high fitness")}");
            }
        }
    }

    class Program
    {
        static void Main()
        {
            var model = new ResourceEvolutionModel();

            // 戦略の定義
            model.AddStrategy("Cooperative", 0.3); // 協力：低コスト
            model.AddStrategy("Defective", 0.1); // 裏切り：低コスト、高適応度

            // 状態の定義
            model.AddState("s1", new double[] { 0.5, 0.5 }, 10.0); // 初期分布、リソース豊富
            model.AddState("s2", new double[] { 0.8, 0.2 }, 5.0); // 協力優勢、リソース中
            model.AddState("s3", new double[] { 0.2, 0.8 }, 1.0); // 裏切り優勢、リソース希少

            // アクセシビリティ関係の定義
            model.AddTransition("s1", "s2", "Cooperative", 0.6);
            model.AddTransition("s1", "s3", "Defective", 0.4);
            model.AddTransition("s2", "s3", "Defective", 0.5);
            model.AddTransition("s2", "s2", "Cooperative", 0.5);
            model.AddTransition("s3", "s3", "Defective", 0.8);
            model.AddTransition("s3", "s2", "Cooperative", 0.2);

            // シミュレーション実行
            model.RunSimulation();
        }
    }
}
